File: idea/src/org/jetbrains/kotlin/idea/projectView/KtDeclarationTreeNode.java
/*
 * Copyright 2010-2015 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.jetbrains.kotlin.idea.projectView;

import com.intellij.application.options.CodeStyle;
import com.intellij.ide.projectView.PresentationData;
import com.intellij.ide.projectView.ViewSettings;
import com.intellij.ide.projectView.impl.nodes.AbstractPsiBasedNode;
import com.intellij.ide.util.treeView.AbstractTreeNode;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiElement;
import org.jetbrains.kotlin.idea.core.formatter.KotlinCodeStyleSettings;
import org.jetbrains.kotlin.psi.*;

import java.util.Collection;
import java.util.Collections;
import java.util.List;

public class KtDeclarationTreeNode extends AbstractPsiBasedNode<KtDeclaration> {
    public static final String CLASS_INITIALIZER = "<class initializer>";

    protected KtDeclarationTreeNode(Project project, KtDeclaration ktDeclaration, ViewSettings viewSettings) {
        super(project, ktDeclaration, viewSettings);
    }

    @Override
    protected PsiElement extractPsiFromValue() {
        return getValue();
    }

    @Override
    protected Collection<AbstractTreeNode<?>> getChildrenImpl() {
        return Collections.emptyList();
    }

    @Override
    protected void updateImpl(PresentationData data) {
        KtDeclaration declaration = getValue();
        if (declaration != null) {
            String text = declaration instanceof KtAnonymousInitializer ? CLASS_INITIALIZER : declaration.getName();
            if (text == null) return;

            KotlinCodeStyleSettings settings = CodeStyle.getSettings(getProject())
                    .getCustomSettings(KotlinCodeStyleSettings.class);

            if (declaration instanceof KtProperty) {
                KtProperty property = (KtProperty) declaration;
                KtTypeReference ref = property.getTypeReference();
                if (ref != null) {
                    if (settings.SPACE_BEFORE_TYPE_COLON) text += " ";
                    text += ":";
                    if (settings.SPACE_AFTER_TYPE_COLON) text += " ";
                    text += ref.getText();
                }
            }
            else if (declaration instanceof KtFunction) {
                KtFunction function = (KtFunction) declaration;
                KtTypeReference receiverTypeRef = function.getReceiverTypeReference();
                if (receiverTypeRef != null) {
                    text = receiverTypeRef.getText() + "." + text;
                }
                text += "(";
                List<KtParameter> parameters = function.getValueParameters();
                for (KtParameter parameter : parameters) {
                    if (parameter.getName() != null) {
                        text += parameter.getName();
                        if (settings.SPACE_BEFORE_TYPE_COLON) text += " ";
                        text += ":";
                        if (settings.SPACE_AFTER_TYPE_COLON) text += " ";
                    }
                    KtTypeReference typeReference = parameter.getTypeReference();
                    if (typeReference != null) {
                        text += typeReference.getText();
                    }
                    text += ", ";
                }
                if (parameters.size() > 0) text = text.substring(0, text.length() - 2);
                text += ")";
                KtTypeReference typeReference = function.getTypeReference();
                if (typeReference != null) {
                    if (settings.SPACE_BEFORE_TYPE_COLON) text += " ";
                    text += ":";
                    if (settings.SPACE_AFTER_TYPE_COLON) text += " ";
                    text += typeReference.getText();
                }
            }

            data.setPresentableText(text);
        }
    }

    @Override
    protected boolean isDeprecated() {
        return KtPsiUtil.isDeprecated(getValue());
    }
}

File: idea/src/org/jetbrains/kotlin/idea/projectView/KtDeclarationTreeNode.kt
/*
 * Copyright 2010-2020 JetBrains s.r.o. and Kotlin Programming Language contributors.
 * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
 */
package org.jetbrains.kotlin.idea.projectView

import com.intellij.application.options.CodeStyle
import com.intellij.ide.projectView.PresentationData
import com.intellij.ide.projectView.ViewSettings
import com.intellij.ide.projectView.impl.nodes.AbstractPsiBasedNode
import com.intellij.ide.util.treeView.AbstractTreeNode
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiElement
import org.jetbrains.kotlin.idea.KotlinBundle
import org.jetbrains.kotlin.idea.core.formatter.KotlinCodeStyleSettings
import org.jetbrains.kotlin.psi.*

internal class KtDeclarationTreeNode(
    project: Project?,
    ktDeclaration: KtDeclaration?,
    viewSettings: ViewSettings?
) : AbstractPsiBasedNode<KtDeclaration?>(project, ktDeclaration!!, viewSettings) {

    override fun extractPsiFromValue(): PsiElement? = value

    override fun getChildrenImpl(): Collection<AbstractTreeNode<*>> = emptyList()

    override fun updateImpl(data: PresentationData) {
        val declaration = value ?: return
        val project = project ?: return
        data.presentableText = tryGetRepresentableText(declaration, project)
    }

    override fun isDeprecated(): Boolean = value?.let { KtPsiUtil.isDeprecated(it) } ?: false

    companion object {
        private val CLASS_INITIALIZER = "<" + KotlinBundle.message("project.view.class.initializer") + ">"

        fun tryGetRepresentableText(declaration: KtDeclaration, project: Project): String? {

            val settings = CodeStyle.getSettings(project).getCustomSettings(KotlinCodeStyleSettings::class.java)
            fun StringBuilder.appendColon() {
                if (settings.SPACE_BEFORE_TYPE_COLON) append(" ")
                append(":")
                if (settings.SPACE_AFTER_TYPE_COLON) append(" ")
            }

            fun KtProperty.presentableText() = buildString {
                append(name)
                typeReference?.text?.let { reference ->
                    appendColon()
                    append(reference)
                }
            }

            fun KtFunction.presentableText() = buildString {
                receiverTypeReference?.text?.let { receiverReference ->
                    append(receiverReference)
                    append('.')
                }
                append(name)
                append("(")
                val valueParameters = valueParameters
                valueParameters.forEachIndexed { index, parameter ->
                    parameter.name?.let { parameterName ->
                        append(parameterName)
                        appendColon()
                    }
                    parameter.typeReference?.text?.let { typeReference ->
                        append(typeReference)
                    }
                    if (index != valueParameters.size - 1) {
                        append(", ")
                    }
                }
                append(")")

                typeReference?.text?.let { returnTypeReference ->
                    appendColon()
                    append(returnTypeReference)
                }
            }

            fun KtObjectDeclaration.presentableText(): String? = buildString {

                if (isCompanion()) {
                    append("companion object")
                } else {
                    append("object")
                    if (!name.isNullOrEmpty()) {
                        append(" $name")
                    }
                }

                if (superTypeListEntries.any()) {
                    appendColon()
                    val lastIndex = superTypeListEntries.size - 1
                    superTypeListEntries.forEachIndexed { index, entry ->
                        entry.typeReference?.text?.let {
                            if (index > 0 && index != lastIndex) append(", ")
                            append(it)
                        }
                    }
                }
            }

            return when (declaration) {
                is KtProperty -> declaration.presentableText()
                is KtFunction -> declaration.presentableText()
                is KtObjectDeclaration -> declaration.presentableText()
                is KtAnonymousInitializer -> CLASS_INITIALIZER
                else -> declaration.name
            }
        }
    }
}